### PR TITLE
mongodoc: introduce development revisions and development ACLs.

### PR DESCRIPTION
**Pull request by frankban**
Also implement the corresponding migration.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1232/
Test FAILed.

File: internal/charmstore/migrations.go [hunk 1]
```diff
@@ -138,6 +142,36 @@ func addSupportedSeries(db StoreDatabase) error {
 	return nil
 }
 
+// handleDevelopmentEntities adds the Development field to all entities and
```
Comment (mhilton) [inline internal/charmstore/migrations.go]: handle doesn't seem to be the correct verb here perhaps it should be two migrations, addDevelopment & addDevelopmentACLs

File: internal/charmstore/migrations.go [hunk 2]
```diff
@@ -138,6 +142,36 @@ func addSupportedSeries(db StoreDatabase) error {
...
+// sets up ACLs for development revisions.
+func handleDevelopmentEntities(db StoreDatabase) error {
+	logger.Infof("adding development field to all entities")
+	if _, err := db.Entities().UpdateAll(nil, bson.D{{
```
Comment (mhilton) [inline internal/charmstore/migrations.go]: This should probably only update those that don't have a development field already. That way it's idempotent and its fairly easy to do.

File: internal/charmstore/migrations.go [hunk 3]
```diff
@@ -138,6 +142,36 @@ func addSupportedSeries(db StoreDatabase) error {
...
+		"$set", bson.D{{"development", false}},
+	}}); err != nil {
+		return errgo.Notef(err, "cannot add development field to all entities")
+	}
+
+	logger.Infof("adding development ACLs to all base entities")
+	baseEntities := db.BaseEntities()
+	var baseEntity mongodoc.BaseEntity
+	iter := baseEntities.Find(bson.D{{
+		"developmentacls", bson.D{{"$exists", false}},
```
Comment (mhilton) [inline internal/charmstore/migrations.go:169]: It should be possible to do this without retrieving every document locally, but it isn't so... moan

File: internal/charmstore/store_test.go [hunk 4]
```diff
@@ -1119,6 +1126,15 @@ func (s *StoreSuite) TestAddUserOwnedCharmArchive(c *gc.C) {
 	s.checkAddCharm(c, charmArchive, false, newResolvedURL("~charmers/precise/wordpress-1", -1))
 }
 
+func (s *StoreSuite) TestAddDevelopmentCharmArchive(c *gc.C) {
+	charmArchive := storetesting.Charms.CharmArchive(c.MkDir(), "wordpress")
+	url := newResolvedURL("~charmers/precise/wordpress-1", 1)
```
Comment (mhilton) [inline internal/charmstore/store_test.go:1131]: perhaps add a boolean to newResolvedURL for development as that function is only used for tests, although it's used in lots of them.

File: internal/mongodoc/doc.go [hunk 5]
```diff
@@ -146,11 +151,13 @@ type BaseEntity struct {
...
 	Public bool
 
-	// ACLs holds permission information relevant to
-	// the base entity. The permissions apply to all
-	// revisions.
+	// ACLs holds permission information relevant to the base entity.
+	// The permissions apply to all revisions.
 	ACLs ACL
 
+	// DevelopmentACLs is similar to ACLs but applies to development revisions.
```
Comment (mhilton) [inline internal/mongodoc/doc.go]: all development revisions?

**Comment by mhilton** [comment]
LGTM with a few thoughts.

**Comment by bac** [comment]
Hi Francesco, the feature branch problem with CI is that it was using 'make build' but the Makefile has:

PROJECT := gopkg.in/juju/charmstore.v5-unstable
PROJECT_DIR := $(shell go list -e -f '{{.Dir}}' $(PROJECT))

Since the default branch 'v5-unstable' is hard-coded in the Makefile using it for feature branches does not work.

An easy fix would be to have 'go list' compare to the project parent:

PROJECT_DIR := $(shell go list -e -f '{{.Dir}}' $(PROJECT)/..)

That symbol is only used to test to ensure the cwd is on the GOPATH so it would not affect anything else.  Make this change as a drive-by if you want.  Otherwise the Makefile is useless for any work outside of the hard-code branch.

**Comment by bac** [comment]
Another problem.  v5-unstable is hardcoded in imports like:

```
"gopkg.in/juju/charmstore.v5-unstable/elasticsearch"
```

File: internal/charmstore/migrations.go (hunk 1 above)
Comment (frankban) [inline internal/charmstore/migrations.go]: Done.

File: internal/charmstore/migrations.go (hunk 2 above)
Comment (frankban) [inline internal/charmstore/migrations.go]: Good idea, done.

File: internal/charmstore/migrations.go (hunk 3 above)
Comment (frankban) [inline internal/charmstore/migrations.go:169]: I am not aware of a way to do that.

File: internal/charmstore/store_test.go (hunk 4 above)
Comment (frankban) [inline internal/charmstore/store_test.go:1131]: I'd prefer to wait a bit for the proper fix and not change the function signature and therefore the function calls everywhere, just to change them back later.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1241/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1243/
Test PASSed.

File: internal/charmstore/migrations.go (hunk 3 above)
Comment (mhilton) [inline internal/charmstore/migrations.go:169]: it's not obviously possible with mongodb

File: internal/charmstore/store_test.go (hunk 4 above)
Comment (mhilton) [inline internal/charmstore/store_test.go:1131]: sounds good

**Comment by urosj** [comment]
:+1: 

QA: migration passed, also, result of this was empty:

db.base_entities.aggregate([{
    $project : {
        equal : {
            $eq : ["$acls", "$developmentacls"] 
        },
        doc : "$$ROOT"  
    }
}, {
    $match : {
        equal : false  
    }
}]);

**Comment by frankban** [comment]
:shipit: 
Thanks Martin and Uros for the reviews!

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**